File: src/types/Race.ts
import * as DB from "tsondb/schema/dsl"
import { CommonnessRatedAdvantageDisadvantage } from "./_CommonnessRatedAdvantageDisadvantage.js"
import { Dice, DieType } from "./_Dice.js"
import {
  AdvantageIdentifier,
  AttributeIdentifier,
  CultureIdentifier,
  DisadvantageIdentifier,
  EyeColorIdentifier,
  HairColorIdentifier,
  RaceIdentifier,
} from "./_Identifier.js"
import { ExperienceLevel } from "./ExperienceLevel.js"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { src } from "./source/_PublicationRef.js"

export const Race = DB.Entity(import.meta.url, {
  name: "Race",
  namePlural: "Races",
  type: () =>
    DB.Object({
      ap_value: DB.Required({
        comment: "How many Adventure Points does the race cost during character creation?",
        type: DB.Integer({ minimum: 0 }),
      }),
      base_values: DB.Required({
        comment: "The race’s base values.",
        type: DB.IncludeIdentifier(BaseValues),
      }),
      attribute_adjustments: DB.Required({
        comment:
          "Describes how to raise or lower maximum attribute values during character creation.",
        type: DB.IncludeIdentifier(AttributeAdjustments),
      }),
      automatic_advantages: DB.Optional({
        comment:
          "A list of automatically applied advantages. This does only work for advantages with no further configuration such as level or special selection.",
        type: DB.Array(
          DB.GenIncludeIdentifier(AutomaticAdvantageDisadvantage, [AdvantageIdentifier()]),
          {
            minItems: 1,
          },
        ),
      }),
      automatic_disadvantages: DB.Optional({
        comment:
          "A list of automatically applied disadvantages. This does only work for disadvantages with no further configuration such as level or special selection.",
        type: DB.Array(
          DB.GenIncludeIdentifier(AutomaticAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      strongly_recommended_advantages: DB.Optional({
        comment: "A list of strongly recommended advantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [AdvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      strongly_recommended_disadvantages: DB.Optional({
        comment: "A list of strongly recommended disadvantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      weight: DB.Required({
        comment: "Configuration for random weight generation.",
        type: DB.IncludeIdentifier(RandomWeightGeneration),
      }),
      starting_age: DB.Required({
        comment:
          "Defines the starting ages for the race. It depends on the selected experience level.",
        type: DB.NestedEntityMap({
          name: "StartingAgeConfigForExperienceLevel",
          namePlural: "StartingAgeConfigsForExperienceLevel",
          secondaryEntity: ExperienceLevel,
          type: DB.Object({
            base: DB.Required({
              comment: "The base value for the selected experience level.",
              type: DB.Integer({ minimum: 1 }),
            }),
            random: DB.Required({
              comment:
                "The random value for the selected experience level. It is going to be added to the base value.",
              type: DB.IncludeIdentifier(Dice),
            }),
          }),
        }),
      }),
      variants: DB.Required({
        comment:
          "A list of available race variants where one has to be selected. If no variants are to be selected, a single variant with no name has to be provided which will be used as the missing values for the base race.",
        type: DB.ChildEntities(RaceVariant),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "Race",
        DB.Object({
          name: DB.Required({
            comment: "The race’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          attribute_adjustments: DB.Required({
            comment: "The respective attribute adjustments text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          automatic_advantages: DB.Optional({
            comment: "The respective automatic advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          automatic_disadvantages: DB.Optional({
            comment: "The respective automatic disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          strongly_recommended_advantages: DB.Optional({
            comment: "The respective strongly recommended advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          strongly_recommended_disadvantages: DB.Optional({
            comment: "The respective strongly recommended disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const BaseValues = DB.TypeAlias(import.meta.url, {
  name: "BaseValues",
  type: () =>
    DB.Object({
      life_points: DB.Required({
        comment: "The race’s life point base value.",
        type: DB.Integer(),
      }),
      spirit: DB.Required({
        comment: "The race’s Spirit base value.",
        type: DB.Integer(),
      }),
      toughness: DB.Required({
        comment: "The race’s Toughness base value.",
        type: DB.Integer(),
      }),
      movement: DB.Required({
        comment: "The race’s tactical movement rate.",
        type: DB.Integer(),
      }),
    }),
})

const AttributeAdjustments = DB.TypeAlias(import.meta.url, {
  name: "AttributeAdjustments",
  comment: "Describes how to raise or lower maximum attribute values during character creation.",
  type: () =>
    DB.Object(
      {
        fixed: DB.Optional({
          comment: "The values by which the maximum of the respective attribute is modified.",
          type: DB.Array(DB.IncludeIdentifier(FixedAttributeAdjustment), { minItems: 1 }),
        }),
        selectable: DB.Optional({
          comment: `An array of attribute maximum modifiers, where the attribute they apply to is selected from a list of options.

The array only permits a single entry because no race specified more than one selectable attribute adjustment so far. But the schema allows for multiple entries to be future-proof.`,
          type: DB.Array(DB.IncludeIdentifier(SelectableAttributeAdjustment), {
            minItems: 1,
            maxItems: 1,
          }),
        }),
      },
      { minProperties: 1 },
    ),
})

const FixedAttributeAdjustment = DB.TypeAlias(import.meta.url, {
  name: "FixedAttributeAdjustment",
  comment: "A value by which the maximum of the respective attribute is modified.",
  type: () =>
    DB.Object({
      id: DB.Required({
        comment: "The attribute the modifier applies to.",
        type: AttributeIdentifier(),
      }),
      value: DB.Required({
        comment:
          "The value by which the specified attribute’s maximum is modified (negative values will lower the maximum).",
        type: DB.Integer(),
      }),
    }),
})

const SelectableAttributeAdjustment = DB.TypeAlias(import.meta.url, {
  name: "SelectableAttributeAdjustment",
  comment:
    "A value that will be added to the current maximum of the selected attribute that has been chosen from the listed attributes (negative values will lower the maximum).",
  type: () =>
    DB.Object({
      list: DB.Required({
        comment: "A list of attributes the player has to choose from.",
        type: DB.Array(AttributeIdentifier(), { minItems: 2 }),
      }),
      value: DB.Required({
        comment:
          "The value by which the selected attribute’s maximum is modified (negative values will lower the maximum).",
        type: DB.Integer(),
      }),
    }),
})

const AutomaticAdvantageDisadvantage = DB.GenTypeAlias(import.meta.url, {
  name: "AutomaticAdvantageDisadvantage",
  comment:
    "An advantage or disadvantage that is automatically applied to the character after selecting the race. This does only work for advantages or disadvantages with no further configuration such as level or special selection.",
  parameters: [DB.Param("Identifier")],
  type: Identifier =>
    DB.Object({
      id: DB.Required({
        comment: "The automatic advantage or disadvantage.",
        type: DB.TypeArgument(Identifier),
      }),
    }),
})

const RandomWeightGeneration = DB.TypeAlias(import.meta.url, {
  name: "RandomWeightGeneration",
  comment: "Configuration for random weight generation.",
  type: () =>
    DB.Object({
      base: DB.Required({
        comment:
          "The base value used for random weight. The height subtrahend; in case of `Height - 110 + 2D6` it is `110`.",
        type: DB.Integer({ minimum: 1 }),
      }),
      random: DB.Required({
        comment: "The dice used for random weight.",
        type: DB.Array(DB.IncludeIdentifier(WeightDice), { minItems: 1 }),
      }),
    }),
})

const WeightDice = DB.TypeAlias(import.meta.url, {
  name: "WeightDice",
  type: () =>
    DB.Object({
      number: DB.Required({
        comment: "Number of dice of the same type. Example: 2 in 2D6.",
        type: DB.Integer({ minimum: 1 }),
      }),
      sides: DB.Required({
        comment: "Number of sides on every die. Example: 6 in 2D6.",
        type: DB.IncludeIdentifier(DieType),
      }),
      offset_strategy: DB.Required({
        comment:
          "The strategy how to offset the randomly generated values against the base value. Either they are all added or subtracted or even results are added and odd results are subtracted.",
        type: DB.IncludeIdentifier(WeightDiceOffsetStrategy),
      }),
    }),
})

const WeightDiceOffsetStrategy = DB.Enum(import.meta.url, {
  name: "WeightDiceOffsetStrategy",
  comment:
    "The strategy how to offset the randomly generated values against the base value. Either they are all added or subtracted or even results are added and odd results are subtracted.",
  values: () => ({
    Add: DB.EnumCase({ type: null }),
    Subtract: DB.EnumCase({ type: null }),
    AddEvenSubtractOdd: DB.EnumCase({ type: null }),
  }),
})

export const RaceVariant = DB.Entity(import.meta.url, {
  name: "RaceVariant",
  namePlural: "RaceVariants",
  type: () =>
    DB.Object({
      race: DB.Required({
        comment: "The associated race.",
        type: RaceIdentifier(),
      }),
      common_cultures: DB.Optional({
        comment: "The list of common cultures.",
        type: DB.Array(CultureIdentifier(), { minItems: 1 }),
      }),
      common_advantages: DB.Optional({
        comment: "A list of common advantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [AdvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      common_disadvantages: DB.Optional({
        comment: "A list of common disadvantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      uncommon_advantages: DB.Optional({
        comment: "A list of uncommon advantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [AdvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      uncommon_disadvantages: DB.Optional({
        comment: "A list of uncommon disadvantages.",
        type: DB.Array(
          DB.GenIncludeIdentifier(CommonnessRatedAdvantageDisadvantage, [DisadvantageIdentifier()]),
          { minItems: 1 },
        ),
      }),
      hair_color: DB.Optional({
        comment:
          "An array containing 20 (numeric) hair color identifiers. The array also represents the 20-sided die for a random hair color.",
        type: DB.Array(HairColorIdentifier(), { minItems: 20, maxItems: 20 }),
      }),
      eye_color: DB.Optional({
        comment:
          "An array containing 20 (numeric) eye color identifiers. The array also represents the 20-sided die for a random eye color.",
        type: DB.Array(EyeColorIdentifier(), { minItems: 20, maxItems: 20 }),
      }),
      height: DB.Required({
        comment: "Configuration for random height generation.",
        type: DB.IncludeIdentifier(RandomHeightGeneration),
      }),
      translations: NestedTranslationMap(
        DB.Required,
        "RaceVariant",
        DB.Object({
          name: DB.Required({
            comment:
              "The race variant’s name. If this is the only variant for a base race and thus just provides the missing information without actually being able to select, fill in the name of the base race.",
            type: DB.String({ minLength: 1 }),
          }),
          common_advantages: DB.Optional({
            comment: "The respective common advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          common_disadvantages: DB.Optional({
            comment: "The respective common disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          uncommon_advantages: DB.Optional({
            comment: "The respective uncommon advantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
          uncommon_disadvantages: DB.Optional({
            comment: "The respective uncommon disadvantages text from the source book.",
            type: DB.String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  parentReferenceKey: "race",
  instanceDisplayName: {},
  uniqueConstraints: [
    [
      {
        keyPath: "race",
      },
      {
        entityMapKeyPath: "translations",
        keyPathInEntityMap: "name",
      },
    ],
  ],
})

const RandomHeightGeneration = DB.TypeAlias(import.meta.url, {
  name: "RandomHeightGeneration",
  comment: "Configuration for random height generation.",
  type: () =>
    DB.Object({
      base: DB.Required({
        comment: "The base value used for random height.",
        type: DB.Integer({ minimum: 1 }),
      }),
      random: DB.Required({
        comment: "The dice used for random height.",
        type: DB.Array(DB.IncludeIdentifier(Dice), { minItems: 1 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/Race.swift
import FileDB

@Model
public struct Race {
    /// How many Adventure Points does the race cost during character creation?
    @Minimum(0)
    let ap_value: Int

    /// The race’s base values.
    let base_values: BaseValues

    /// Describes how to raise or lower maximum attribute values during character creation.
    let attribute_adjustments: AttributeAdjustments

    /// A list of automatically applied advantages. This does only work for advantages with no further configuration such as level or special selection.
    @MinItems(1)
    @Relationship(Advantage.self)
    let automatic_advantages: [Advantage.ID]?

    /// A list of strongly recommended advantages.
    @MinItems(1)
    @Relationship(Advantage.self)
    let strongly_recommended_advantages: [CommonnessRatedAdvantageDisadvantage<Advantage.ID>]?

    /// A list of strongly recommended disadvantages.
    @MinItems(1)
    @Relationship(Disadvantage.self)
    let strongly_recommended_disadvantages: [CommonnessRatedAdvantageDisadvantage<Disadvantage.ID>]?

    /// Configuration for random weight generation.
    let weight: RandomWeightGeneration

    /// Defines the starting ages for the race. It depends on the selected experience level.
    @Relationship(ExperienceLevel.self)
    let starting_age: [Locale.ID: StartingAge]

    /// A list of available race variants where one has to be selected. If no variants are to be selected, a single variant with no name has to be provided which will be used as the missing values for the base race.
    @Relationship(RaceVariant.self)
    @MinItems(1)
    let variants: [RaceVariant.ID]
    // isTransient: true,

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // RaceTranslation
        /// The race’s name.
        @MinLength(1)
        let name: String

        /// The respective attribute adjustments text from the source book.
        @MinLength(1)
        let attribute_adjustments: String

        /// The respective automatic advantages text from the source book.
        @MinLength(1)
        let automatic_advantages: String

        /// The respective strongly recommended advantages text from the source book.
        @MinLength(1)
        let strongly_recommended_advantages: String

        /// The respective strongly recommended disadvantages text from the source book.
        @MinLength(1)
        let strongly_recommended_disadvantages: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

@Embedded
public struct BaseValues {
    /// The race’s life point base value.
    let life_points: Int

    /// The race’s Spirit base value.
    let spirit: Int

    /// The race’s Toughness base value.
    let toughness: Int

    /// The race’s tactical movement rate.
    let movement: Int
}

/// Describes how to raise or lower maximum attribute values during character creation.
@Embedded
@MinProperties(1)
public struct AttributeAdjustments {
    /// The values by which the maximum of the respective attribute is modified.
    @MinItems(1)
    let fixed: [FixedAttributeAdjustment]?

    /// An array of attribute maximum modifiers, where the attribute they apply to is selected from a list of options.
    ///
    /// The array only permits a single entry because no race specified more than one selectable attribute adjustment so far. But the schema allows for multiple entries to be future-proof.
    @MinItems(1)
    @MaxItems(1)
    let selectable: [SelectableAttributeAdjustment]?
}

/// A value by which the maximum of the respective attribute is modified.
@Embedded
public struct FixedAttributeAdjustment {
    /// The attribute the modifier applies to.
    @Relationship(Attribute.self)
    let id: Attribute.ID

    /// The value by which the specified attribute’s maximum is modified (negative values will lower the maximum).
    let value: Int
}

/// A value that will be added to the current maximum of the selected attribute that has been chosen from the listed attributes (negative values will lower the maximum).
@Embedded
public struct SelectableAttributeAdjustment {
    /// A list of attributes the player has to choose from.
    @MinItems(2)
    @Relationship(Attribute.self)
    let list: [Attribute.ID]

    /// The value by which the selected attribute’s maximum is modified (negative values will lower the maximum).
    let value: Int
}

/// Configuration for random weight generation.
@Embedded
public struct RandomWeightGeneration {
    /// The base value used for random weight. The height subtrahend; in case of `Height - 110 + 2D6` it is `110`.
    @Minimum(1)
    let base: Int

    /// The dice used for random weight.
    @MinItems(1)
    let random: [WeightDice]
}

@Embedded
public struct WeightDice {
    /// Number of dice of the same type. Example: 2 in 2D6.
    @Minimum(1)
    let number: Int

    /// Number of sides on every die. Example: 6 in 2D6.
    let sides: DieType

    /// The strategy how to offset the randomly generated values against the base value. Either they are all added or subtracted or even results are added and odd results are subtracted.
    let offset_strategy: WeightDiceOffsetStrategy
}

/// The strategy how to offset the randomly generated values against the base value. Either they are all added or subtracted or even results are added and odd results are subtracted.
@ModelEnum
public enum WeightDiceOffsetStrategy {
    case add
    case subtract
    case addEvenSubtractOdd
}

@Model
public struct RaceVariant {
    /// The associated race.
    @Relationship(Race.self)
    let race: Race.ID

    /// The list of common cultures.
    @MinItems(1)
    @Relationship(Culture.self)
    let common_cultures: [Culture.ID]?

    /// A list of common advantages.
    @Relationship(Advantage.self)
    @MinItems(1)
    let common_advantages: [CommonnessRatedAdvantageDisadvantage<Advantage.ID>]?

    /// A list of common disadvantages.
    @Relationship(Disadvantage.self)
    @MinItems(1)
    let common_disadvantages: [CommonnessRatedAdvantageDisadvantage<Disadvantage.ID>]?

    /// A list of uncommon advantages.
    @Relationship(Advantage.self)
    @MinItems(1)
    let uncommon_advantages: [CommonnessRatedAdvantageDisadvantage<Advantage.ID>]?

    /// A list of uncommon disadvantages.
    @Relationship(Disadvantage.self)
    @MinItems(1)
    let uncommon_disadvantages: [CommonnessRatedAdvantageDisadvantage<Disadvantage.ID>]?

    /// An array containing 20 (numeric) hair color identifiers. The array also represents the 20-sided die for a random hair color.
    @MinItems(20)
    @MaxItems(20)
    @Relationship(HairColor.self)
    let hair_color: [HairColor.ID]?

    /// An array containing 20 (numeric) eye color identifiers. The array also represents the 20-sided die for a random eye color.
    @MinItems(20)
    @MaxItems(20)
    @Relationship(EyeColor.self)
    let eye_color: [EyeColor.ID]?

    /// Configuration for random height generation.
    let height: RandomHeightGeneration

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // RaceVariantTranslation
        /// The race variant’s name. If this is the only variant for a base race and thus just provides the missing information without actually being able to select, fill in the name of the base race.
        @MinLength(1)
        let name: String

        /// The respective common advantages text from the source book.
        @MinLength(1)
        let common_advantages: String?

        /// The respective common disadvantages text from the source book.
        @MinLength(1)
        let common_disadvantages: String?

        /// The respective uncommon advantages text from the source book.
        @MinLength(1)
        let uncommon_advantages: String?

        /// The respective uncommon disadvantages text from the source book.
        @MinLength(1)
        let uncommon_disadvantages: String?
    }
}

/// Configuration for random height generation.
@Embedded
public struct RandomHeightGeneration {
    /// The base value used for random height.
    @Minimum(1)
    let base: Int

    /// The dice used for random height.
    @MinItems(1)
    let random: [Dice]
}

@Embedded
struct StartingAge {
    /// The base value for the selected experience level.
    @Minimum(1)
    let base: Int

    /// The random value for the selected experience level. It is going to be added to the base value.
    let random: Dice
}
